### ISSUE TITLE
[Python] Read table stuck and hangs forever

### ISSUE THREAD

**Issue by unReaLaneS** [apache/arrow#37139]
### Describe the bug, including details regarding any error messages, version, and platform.

File is parquet file and it is being read from the s3 directly.
The parquet file size is 1.4 GB.

Here is the code:
for batch in pq.read_table("bucket_path", filesystem=self.s3_file_system).to_batches(batch_size)

It stucks and there is no exception or anything.

### Component(s)

Parquet, Python

**Comment by mapleFU** [comment]
Would you mind set log level and see the logging of s3? I don't know how time spends here, need more log to see why initialize is slow?

**Comment by unReaLaneS** [comment]
Hi, thanks for answering. I checked the memory usage, it seems pq.read_table().to_batches() loads entire file into memory, is it supposed to do so, cause it can be misunderstood that to_batches is going to take partial data. I also tracked locally and on the server, it is fetching entire file and then, until next file comes it is not calling s3 at all.

Solution that I found for this issue, is to switch to the:
`pq.ParquetFile("bucket_path", filesystem=self.s3_file_system).iter_batches(batch_size)`, which indeed loads data in batches into memory.

**Comment by mapleFU** [comment]
Hmmm as for behavior, parquet will usally load S3 in the granularity of parquet column chunkes, but `iter_batches` or other method might "read-ahead" to prebuffer the file or row-groups

Maybe I can latter add to document for that

**Comment by unReaLaneS** [comment]
Thank you, very much <3

**Comment by unReaLaneS** [comment]
It seems, it is a problem with memory leak, when I get batches in to_batches or iter_batches and loop through generator of RecordBatch, it is taking up memory on every iteration and it is not getting freed.

Here is the memory graph after loading 198 GB of data(compressed in multiple parquet files) in AWS FargateTask, it is being killed after it runs out of memory. I tried with gc.collect() and every optimization, deleting object, but still it keeps going up, even when I switch to load other file, the memory taken by previous file, it is still allocated.

![image](https://github.com/apache/arrow/assets/7562818/1fb89fb6-3ee7-4375-8c5a-42b0bfc57ac9)

**Comment by mapleFU** [comment]
I guess it's also because prefetching. Prefetch will read more object to memory from s3. It prefetch depth is high, it will prefetch lots of files

**Comment by unReaLaneS** [comment]
Yeah, I get it that it is doing prefetching, but why memory is not dropping after iteration, with batch, but it is increasing after every iteration, is it possible cause of the Python and Linux combination? I also noticed in the macOS, it is being steady, but not quite sure, this happens over the span of 6 hours.

**Comment by mapleFU** [comment]
I'm just back from business trip, sorry for late reply

`iter_batches` is a function which frequently "yield" batches, it would create a RecordBatchReader, and continue "yield" the contents. How can I reproduce your scenerio? I'm not sure where cost the amount of memory

**Comment by mapleFU** [comment]
> Yeah, I get it that it is doing prefetching, but why memory is not dropping after iteration, with batch, but it is increasing after every iteration, is it possible cause of the Python and Linux combination? I also noticed in the macOS, it is being steady, but not quite sure, this happens over the span of 6 hours.

Have you fixed this? I've checked the code and I think memory would be released if fetch is done. If you still meet the problem, would you mind provide the code? And I can debug here. @unReaLaneS

**Comment by mapleFU** [comment]
https://github.com/apache/arrow/issues/37630

This might be a same question

**Comment by unReaLaneS** [comment]
@mapleFU Hi, sorry for taking too long. I manage to increase the RAM and it works now, but still the RAM usage is somewhat linear, so the parquet files are read from the S3, so I solved it by increasing the RAM.

But what I noticed regarding `iter_batches` and `to_batches` is they behave differently on different OS environments.

By the documentation both `iter_batches` and `to_batches` should "yield the batches, but the to_batches also has read ahead buffer mechanism, I get it.

**macOS**:
I was checking memory usage directly in the Activity Monitor. It seems that `iter_batches` behaves like it should behave and for some time memory usage is normal, but overtime it seems to slightly increase, but with `to_batches`, it prebuffers entire parquet file, there is no limitation, if it is 1GB parquet file, it will load it into 16 GB of RAM and then it start doing memory swap.

**Linux**
On the other hand `iter_batches` seems to load entire file and it immediatelly fails due to RAM usage, but on the other hand `to_batches` is loading in batches, even with prebuffering and memory usage is oK!, but overtime it is increasing until the container is killed by AWS.

Here is the code:

```
    def get_all_s3_files(self):
        self.logger.info("Bucket name: " + os.environ.get('RP_BUCKET_NAME'))
        self.logger.info("Bucket resource path: " + os.environ.get('RP_BUCKET_RESOURCE_PATH'))

        self.logger.info("Return list of s3 parquet files")

        for obj in self.my_bucket.objects.filter(Prefix=os.environ.get('RP_BUCKET_RESOURCE_PATH')):
            if obj.key.endswith('parquet'):
                yield obj



    def read_from_s3(self, resource_path, batch_size=int(os.environ.get('READ_BULK_SIZE_FROM_PARQUET'))):

        self.logger.info(f"Bucket name: {resource_path.bucket_name}")
        self.logger.info(f"Bucket full path: {resource_path.bucket_name}/{resource_path.key}")
        self.logger.info(f"Loading from s3 file by batch size: {os.environ.get('READ_BULK_SIZE_FROM_PARQUET')}")

        for batch in pq.read_table(f"{resource_path.bucket_name}/{resource_path.key}", filesystem=self.s3_file_system)\
                .to_batches(batch_size):
            yield json.loads(batch.to_pandas().to_json(orient="records"))
            del batch
            gc.collect()
```

I am reading files from S3 in the loop, so it is read file by file.

```
            for item in s3_client.get_all_s3_files():

                self.logger.info(f'Getting chunks from new file with key {item.key}')

                response["no_of_parquet_files"] = response["no_of_parquet_files"] + 1

                start_time = time.time()

                for batch in s3_client.read_from_s3(item):
```
Here is the code how it is being called. I thought also that It might be the issue with Python and Linux not releasing the memory, but it happens also in the macOS. That issue #37630 might be also the problem, I was also checking that It might be the problem with to_pandas, but there is no other way to transfer data to json, only like this, cause if any other approach is used, data is not being parsed correctly and totally wrong data is returned.

Thanks again, @mapleFU

**Comment by chriss1245** [comment]
Something similar happens in my case. I am using iter_batches from a ParquetFile in order to create a generator for tensorflow. The loop is quite simple. 

```python
for batch in data.iter_batches(columns=columns, batch_size=batch_size, use_threads=False):
        batch_df = batch.to_pandas()
        yield {
            col: tf.convert_to_tensor(batch_df[col].values)
            for col in batch_df.columns
        }
``` 

I created a RAM monitoring callback which monitors the ram after each batch. And I have seen the same behavior: the ram usage increases linearly.  I tried different variations but the pattern is the same. 
![image](https://github.com/apache/arrow/assets/58918297/7a1f04c2-0c72-4725-97ea-e49a434ed5d7)

**Comment by mapleFU** [comment]
I think this share the same reason with https://github.com/apache/arrow/issues/37630

Let me draft and fixing this

**Comment by mapleFU** [comment]
@jorisvandenbossche would you mind check the `to_pandas` here? I'm not familiar with this part of logic...
